### PR TITLE
rgw: CopyObj returns x-amz-version-id

### PR DESCRIPTION
**Pull request by Rjerk**
... when copying an object into a version-enabled bucket

Signed-off-by: Liu Lan <liulan_yewu@cmss.chinamobile.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -3283,6 +3283,8 @@ void RGWCopyObj_ObjStore_S3::send_partial_response(off_t ofs)
     set_req_state_err(s, op_ret);
     dump_errno(s);
 
+    dump_header_if_nonempty(s, "x-amz-copy-source-version-id", src_object->get_instance());
+    dump_header_if_nonempty(s, "x-amz-version-id", dest_object->get_instance());
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc:3287]: looks correct to me, but it's hard to tell whether `src_object` and `dest_object` are guaranteed to be initialized here. do we need to test them for null first?

cc @dang

Comment (dang) [reply src/rgw/rgw_rest_s3.cc:3287]: I think it's okay.  It's only called from send_response(), or ultimately from execute(), both of which should have src_object adn dest_object initialized.

Comment (cbodley) [reply src/rgw/rgw_rest_s3.cc:3287]: even though `op_ret` may be nonzero here? what if we got ENOENT looking up the src_object?

Comment (dang) [reply src/rgw/rgw_rest_s3.cc:3287]: Good point.  I missed that it's not bailing if op_ret is failure.  Maybe that is enough?  If not, then rgw::sal::RGWObject::empty() will need to be used to check that they're okay.

Comment (cbodley) [reply src/rgw/rgw_rest_s3.cc:3287]: @Rjerk can you please surround these two lines with an `if (op_ret == 0) {` block?

Comment (Rjerk) [reply src/rgw/rgw_rest_s3.cc]: done

**Review by dang — APPROVED** [review]

**Comment by Rjerk** [comment]
Add tests for this pr: https://github.com/ceph/s3-tests/pull/383

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by cbodley** [comment]
seems to be failing s3tests:
```
======================================================================
FAIL: s3tests_boto3.functional.test_s3.test_versioning_copy_obj_version
---------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/cephtest/s3-tests/virtualenv/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/ubuntu/cephtest/s3-tests/s3tests_boto3/functional/test_s3.py", line 8601, in test_versioning_copy_obj_version
    eq(body, contents[i])
AssertionError: 'content-2' != 'content-0'
-------------------- >> begin captured logging << --------------------
```
https://pulpito.ceph.com/hmunjulu-2021-02-04_22:29:42-rgw-wip-hmunjulu01-testing-2021-02-04-1031-distro-basic-smithi/

**Comment by ivancich** [comment]
@Rjerk Would you be willing to rebase?

**Comment by Rjerk** [comment]
jenkins test make check

**Comment by Rjerk** [comment]
jenkins test docs

**Comment by Rjerk** [comment]
jenkins retest this please

**Comment by Rjerk** [comment]
rebase is done. any pieces of advice?

**Comment by mattbenjamin** [comment]
@dang is this still viable?

**Comment by dang** [comment]
jenkins retest this please

**Review by dang — CHANGES_REQUESTED** [review]
This is now failing, src_object needs to be changed to s->src_object

**Pull request closed without merging**